File: app/src/main/java/com/pandatone/kumiwake/sekigime/SekigimeResult.java
package com.pandatone.kumiwake.sekigime;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Point;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.Display;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.Toast;

import com.pandatone.kumiwake.R;
import com.pandatone.kumiwake.customize.CustomDialog;
import com.pandatone.kumiwake.kumiwake.MainActivity;
import com.pandatone.kumiwake.member.Name;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by atsushi_2 on 2016/07/16.
 */
public class SekigimeResult extends AppCompatActivity {

    public static ArrayList<ArrayList<Name>> arrayArrayNormal, arrayArrayNormalMan, arrayArrayNormalWoman;
    public static ArrayList<ArrayList<String>> arrayArrayQuick, arrayArrayQuickMan, arrayArrayQuickWoman;
    public static ArrayList<String> groupArray;
    public static boolean Normalmode, doubleDeploy, fmDeploy;
    public static int square_no, groupNo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        float scale = getResources().getDisplayMetrics().density;
        RelativeLayout reLayout = new RelativeLayout(this);
        RelativeLayout.LayoutParams param1 = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        reLayout.setLayoutParams(param1);
        RelativeLayout.LayoutParams param2 = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        param2.addRule(RelativeLayout.CENTER_IN_PARENT);
        param2.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
        param2.setMargins(param2.leftMargin, param2.topMargin, param2.rightMargin, (int) (12 * scale));
        AdView adView = new AdView(this);
        adView.setAdUnitId("ca-app-pub-2315101868638564/8665451539");
        adView.setAdSize(AdSize.BANNER);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice("BB707E3F7B5413908B2DD12063887489").build();
        adView.loadAd(adRequest);

        groupNo = groupArray.size();
        if (fmDeploy) {
            convertAlternatelyFmArray();
        }
        final DrawTableView draw = new DrawTableView(this);
        WindowManager wm = (WindowManager) getSystemService(WINDOW_SERVICE);
        Display disp = wm.getDefaultDisplay();
        Point size = new Point();
        disp.getSize(size);
        Spinner group_spinner = new Spinner(this);
        LinearLayout layout = new LinearLayout(this);
        Button re_sekigime = new Button(this);
        Button go_home = new Button(this);
        int buttonWidth = (int) (180 * scale);
        int buttonHeight = (int) (50 * scale);
        float centerX = (size.x - buttonWidth) / 2;
        re_sekigime.setText(getText(R.string.re_sekigime));
        go_home.setText(getText(R.string.go_home));
        LinearLayout.LayoutParams llp1 = new LinearLayout.LayoutParams(buttonWidth, buttonHeight);
        LinearLayout.LayoutParams llp2 = new LinearLayout.LayoutParams(buttonWidth, buttonHeight);
        ViewGroup.MarginLayoutParams mlp1 = llp1;
        ViewGroup.MarginLayoutParams mlp2 = llp2;
        mlp1.setMargins(mlp1.leftMargin, (int) (10 * scale), mlp1.rightMargin, (int) (10 * scale));
        re_sekigime.setLayoutParams(mlp1);
        re_sekigime.setTranslationX(centerX);
        mlp2.setMargins(mlp2.leftMargin, (int) (10 * scale), mlp2.rightMargin, (int) (90 * scale));
        go_home.setLayoutParams(mlp2);
        go_home.setTranslationX(centerX);
        re_sekigime.setBackground(ContextCompat.getDrawable(this, R.drawable.simple_orange_button));
        go_home.setBackground(ContextCompat.getDrawable(this, R.drawable.simple_green_button));
        re_sekigime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String message = getString(R.string.re_sekigime_description) + getString(R.string.run_confirmation);
                String title = getString(R.string.re_sekigime_title);
                final CustomDialog customDialog = new CustomDialog();
                customDialog.setTitle(title);
                customDialog.setMessage(message);
                customDialog.mPositiveBtnListener = new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        for (int i = 0; i < groupNo; i++) {
                            if (Normalmode) {
                                Collections.shuffle(arrayArrayNormal.get(i));
                            } else {
                                Collections.shuffle(arrayArrayQuick.get(i));
                            }
                        }
                        if (fmDeploy) {
                            convertAlternatelyFmArray();
                        }
                        DrawTableView.point = 0;
                        draw.reDraw();
                        Toast.makeText(getApplicationContext(), getText(R.string.re_sekigime_finished), Toast.LENGTH_SHORT).show();
                        customDialog.dismiss();
                    }
                };
                customDialog.show(getFragmentManager(), "Btn");
            }
        });
        go_home.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(intent);
            }
        });
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setBackground(ContextCompat.getDrawable(this, R.drawable.sekigime_img));

        ArrayAdapter<String> adapter = new ArrayAdapter<String>(this, R.layout.spinner_layout);
        adapter.setDropDownViewResource(
                android.R.layout.simple_spinner_dropdown_item
        );
        List<String> list = new ArrayList<>(); // 新インスタンスを生成
        for (int j = 0; j < groupArray.size(); j++) {
            list.add(groupArray.get(j));
        }
        adapter.addAll(list);
        group_spinner.setAdapter(adapter);
        Spinner.LayoutParams lp = new Spinner.LayoutParams(buttonWidth, buttonHeight);
        group_spinner.setLayoutParams(lp);
        group_spinner.setTranslationY(15 * scale);
        group_spinner.setTranslationX((float) (centerX * 1.1));
        group_spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position,
                                       long id) {
                DrawTableView.point = 0;
                DrawTableView.position = position;
                draw.reDraw();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }

        });

        ScrollView scrollView = new ScrollView(this);
        reLayout.addView(scrollView);
        reLayout.addView(adView, param2);
        layout.addView(group_spinner);
        layout.addView(draw);
        layout.addView(re_sekigime);
        layout.addView(go_home);
        scrollView.addView(layout);
        setContentView(reLayout);
    }

    public void convertAlternatelyFmArray() {
        float manArrayNo, womanArrayNo, bigger, smaller, addNo, remainder;
        int i, j, k, a, memberSum;
        if (Normalmode) {
            createFmArrayForNormal();
            ArrayList<ArrayList<Name>> smallerArray, biggerArray;
            arrayArrayNormal = new ArrayList<>(groupNo);
            for (int g = 0; g < groupNo; g++) {
                arrayArrayNormal.add(new ArrayList<Name>());
            }
            for (i = 0; i < groupNo; i++) {
                manArrayNo = arrayArrayNormalMan.get(i).size();
                womanArrayNo = arrayArrayNormalWoman.get(i).size();
                if (manArrayNo < womanArrayNo) {
                    bigger = womanArrayNo;
                    smaller = manArrayNo;
                    biggerArray = arrayArrayNormalWoman;
                    smallerArray = arrayArrayNormalMan;
                } else {
                    bigger = manArrayNo;
                    smaller = womanArrayNo;
                    biggerArray = arrayArrayNormalMan;
                    smallerArray = arrayArrayNormalWoman;
                }
                if (smaller != 0) {
                    addNo = bigger / smaller;
                    remainder = (bigger % smaller) - 1;
                    memberSum = 0;
                    for (a = 0; a < addNo / 2; a++) {
                        arrayArrayNormal.get(i).add(biggerArray.get(i).get(memberSum));
                        memberSum++;
                    }
                    arrayArrayNormal.get(i).add(smallerArray.get(i).get(0));
                    for (j = 1; j < smaller; j++) {
                        for (k = 0; k < addNo - 1; k++) {
                            arrayArrayNormal.get(i).add(biggerArray.get(i).get(memberSum));
                            memberSum++;
                        }
                        if (remainder != 0) {
                            arrayArrayNormal.get(i).add(biggerArray.get(i).get(memberSum));
                            memberSum++;
                            remainder--;
                        }
                        arrayArrayNormal.get(i).add(smallerArray.get(i).get(j));
                    }
                    for (; a < addNo; a++) {
                        arrayArrayNormal.get(i).add(biggerArray.get(i).get(memberSum));
                        memberSum++;
                    }
                } else {
                    for (k = 0; k < bigger; k++) {
                        arrayArrayNormal.get(i).add(biggerArray.get(i).get(k));
                    }
                }

            }
        } else {
            createFmArrayForQuick();
            ArrayList<ArrayList<String>> smallerArray, biggerArray;
            arrayArrayQuick = new ArrayList<>(groupNo);
            for (int g = 0; g < groupNo; g++) {
                arrayArrayQuick.add(new ArrayList<String>());
            }
            for (i = 0; i < groupNo; i++) {
                manArrayNo = arrayArrayQuickMan.get(i).size();
                womanArrayNo = arrayArrayQuickWoman.get(i).size();
                if (manArrayNo < womanArrayNo) {
                    bigger = womanArrayNo;
                    smaller = manArrayNo;
                    biggerArray = arrayArrayQuickWoman;
                    smallerArray = arrayArrayQuickMan;
                } else {
                    bigger = manArrayNo;
                    smaller = womanArrayNo;
                    biggerArray = arrayArrayQuickMan;
                    smallerArray = arrayArrayQuickWoman;
                }
                if (smaller != 0) {
                    addNo = bigger / smaller;
                    remainder = (bigger % smaller) - 1;
                    memberSum = 0;
                    for (a = 0; a < addNo / 2; a++) {
                        arrayArrayQuick.get(i).add(biggerArray.get(i).get(memberSum));
                        memberSum++;
                    }
                    arrayArrayQuick.get(i).add(smallerArray.get(i).get(0));
                    for (j = 1; j < smaller; j++) {
                        for (k = 0; k < addNo - 1; k++) {
                            arrayArrayQuick.get(i).add(biggerArray.get(i).get(memberSum));
                            memberSum++;
                        }
                        if (remainder != 0) {
                            arrayArrayQuick.get(i).add(biggerArray.get(i).get(memberSum));
                            memberSum++;
                            remainder--;
                        }
                        arrayArrayQuick.get(i).add(smallerArray.get(i).get(j));
                    }
                    for (; a < addNo; a++) {
                        arrayArrayQuick.get(i).add(biggerArray.get(i).get(memberSum));
                        memberSum++;
                    }

                } else {
                    for (k = 0; k < bigger; k++) {
                        arrayArrayQuick.get(i).add(biggerArray.get(i).get(k));
                    }
                }
            }

        }
    }

    public void createFmArrayForNormal() {
        Name item;
        arrayArrayNormalMan = new ArrayList<>(groupNo);
        arrayArrayNormalWoman = new ArrayList<>(groupNo);
        for (int g = 0; g < groupNo; g++) {
            arrayArrayNormalMan.add(new ArrayList<Name>());
            arrayArrayNormalWoman.add(new ArrayList<Name>());
        }
        for (int i = 0; i < arrayArrayNormal.size(); i++) {
            for (int j = 0; j < arrayArrayNormal.get(i).size(); j++) {
                item = arrayArrayNormal.get(i).get(j);
                if (item.getSex().equals(getText(R.string.man))) {
                    arrayArrayNormalMan.get(i).add(item);
                } else {
                    arrayArrayNormalWoman.get(i).add(item);
                }
            }
        }
    }

    public void createFmArrayForQuick() {
        String item;
        arrayArrayQuickMan = new ArrayList<>(groupNo);
        arrayArrayQuickWoman = new ArrayList<>(groupNo);
        for (int g = 0; g < groupNo; g++) {
            arrayArrayQuickMan.add(new ArrayList<String>());
            arrayArrayQuickWoman.add(new ArrayList<String>());
        }
        for (int i = 0; i < arrayArrayQuick.size(); i++) {
            for (int j = 0; j < arrayArrayQuick.get(i).size(); j++) {
                item = arrayArrayQuick.get(i).get(j);
                if (item.matches(".*" + "♠" + ".*")) {
                    arrayArrayQuickMan.get(i).add(item);
                } else {
                    arrayArrayQuickWoman.get(i).add(item);
                }
            }
        }
    }
}

File: app/src/main/java/com/pandatone/kumiwake/sekigime/SekigimeResult.kt
package com.pandatone.kumiwake.sekigime

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.view.inputmethod.InputMethodManager
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.LinearLayout
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.widget.doAfterTextChanged
import com.google.android.material.tabs.TabLayout
import com.google.android.material.textfield.TextInputLayout
import com.pandatone.kumiwake.*
import com.pandatone.kumiwake.history.HistoryAdapter
import com.pandatone.kumiwake.member.function.Member
import com.pandatone.kumiwake.sekigime.function.DrawAllTable
import com.pandatone.kumiwake.sekigime.function.DrawTableView
import com.pandatone.kumiwake.ui.dialogs.DialogWarehouse
import kotlin.math.round


/**
 * Created by atsushi_2 on 2016/07/16.
 */
class SekigimeResult : AppCompatActivity() {

    private lateinit var draw: DrawTableView
    private var groupNo: Int = 0
    private var teamArrayMan: ArrayList<ArrayList<Member>> = ArrayList()
    private var teamArrayWoman: ArrayList<ArrayList<Member>> = ArrayList()
    private lateinit var tabLayout: TabLayout

    //結果タイトル&コメント
    private var title = ""
    private var comment = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.sekigime_result)

        tabLayout = findViewById(R.id.tabLayout)
        tabLayout.addTab(tabLayout.newTab().setText(R.string.show_detail))
        tabLayout.addTab(tabLayout.newTab().setText(R.string.show_all))
        tabLayout.addOnTabSelectedListener(tabItemSelectedListener)

        PublicMethods.showAd(this)
        DrawTableView.tableNo = 0 //表示するテーブル番号

        groupNo = groupArray!!.size
        if (fmDeploy) {
            convertAlternatelyFmArray()
        }
        drawView(0)
        val groupDropdown = findViewById<View>(R.id.group_dropdown) as AutoCompleteTextView
        groupDropdown.onItemClickListener = OnItemClickListener { _, _, _, _ ->
            val manager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            manager.hideSoftInputFromWindow(groupDropdown.windowToken, 0)
        }
        val adapter = ArrayAdapter<String>(this, R.layout.dropdown_item_layout)
        val list = ArrayList<String>() // 新インスタンスを生成
        for (group in groupArray!!) {
            list.add(group)
        }
        adapter.addAll(list)
        groupDropdown.setAdapter(adapter)
        groupDropdown.hint = list[0]
        groupDropdown.onItemClickListener = OnItemClickListener { _, _, position, _ ->
            drawView(position)
        }

        findViewById<ImageButton>(R.id.edit_result_title).setOnClickListener { editInfoDialog() }
        findViewById<Button>(R.id.re_sekigime).setOnClickListener { onReSekigime() }
        findViewById<Button>(R.id.go_home).setOnClickListener { onGoHome() }
        findViewById<Button>(R.id.share_image).setOnClickListener { onShareImage() }
    }

    //選択表示描画
    private fun drawView(position: Int) {
        DrawTableView.tableNo = position
        draw = DrawTableView(this)
        val resultLayout = findViewById<LinearLayout>(R.id.result_layout)
        resultLayout.removeAllViews()
        resultLayout.addView(draw)
    }

    //全表示
    private fun drawAll(resultLayout: LinearLayout) {
        for (group in groupArray!!.withIndex()) {
            val drawAll = DrawAllTable(this@SekigimeResult, group.index)
            val groupNameView = groupTextView(group.value, group.index)
            resultLayout.addView(groupNameView)
            resultLayout.addView(drawAll)
        }
    }

    //表示形式切り替え
    private val tabItemSelectedListener = object : TabLayout.OnTabSelectedListener {

        override fun onTabSelected(tab: TabLayout.Tab) {
            findViewById<ScrollView>(R.id.result_scroller).fullScroll(ScrollView.FOCUS_UP)
            val resultLayout = findViewById<LinearLayout>(R.id.result_layout)
            resultLayout.removeAllViews()
            val dropdown = findViewById<TextInputLayout>(R.id.group_selector)
            val shareButton = findViewById<Button>(R.id.share_image)

            when (tab.position) {
                0 -> {//テーブルごと
                    dropdown.visibility = View.VISIBLE
                    shareButton.visibility = View.GONE
                    resultLayout.addView(draw)
                }
                1 -> {//全表示
                    dropdown.visibility = View.GONE
                    shareButton.visibility = View.VISIBLE
                    drawAll(resultLayout)
                }
            }
        }

        override fun onTabUnselected(tab: TabLayout.Tab) {}
        override fun onTabReselected(tab: TabLayout.Tab) {}
    }

    //情報変更ダイアログ
    private fun editInfoDialog() {
        val builder = AlertDialog.Builder(this)
        val inflater = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = inflater.inflate(R.layout.edit_result_dialog_layout, findViewById<View>(R.id.info_layout) as ViewGroup?)
        val etTitle = view.findViewById<EditText>(R.id.edit_title)
        etTitle.hint = getString(R.string.sekigime_result)
        if (title != "") etTitle.setText(this.title)
        val etComment = view.findViewById<EditText>(R.id.edit_comment)
        if (comment != "") etComment.setText(this.comment)
        view.findViewById<CheckBox>(R.id.include_title_check).visibility = View.GONE
        view.findViewById<CheckBox>(R.id.include_comment_check).visibility = View.GONE
        builder.setTitle(getString(R.string.add_info))
                .setView(view)
                .setPositiveButton(R.string.change) { _, _ ->
                    changeInfo(etTitle.text.toString(), etComment.text.toString())
                    val tvTitle = findViewById<TextView>(R.id.inner_result_title)
                    if (title != "") {
                        tvTitle.visibility = View.VISIBLE
                        if (StatusHolder.normalMode) {
                            val hsAdapter = HistoryAdapter(this)
                            hsAdapter.updateHistoryState(hsAdapter.latestHistory, title, false)
                        }
                    } else {
                        tvTitle.visibility = View.GONE
                    }
                    val tvComment = findViewById<TextView>(R.id.comment_view)
                    if (comment != "") {
                        tvComment.visibility = View.VISIBLE
                    } else {
                        tvComment.visibility = View.GONE
                    }
                }
                .setNegativeButton(R.string.cancel) { dialog, _ ->
                    dialog.dismiss()
                }
        val dialog = builder.create()
        dialog.show()
    }

    //情報変更
    private fun changeInfo(newTitle: String, newComment: String) {
        title = newTitle
        val tvTitle = findViewById<TextView>(R.id.inner_result_title)
        if (title != "") {
            tvTitle.text = title
        } else {
            tvTitle.text = getString(R.string.sekigime_result)
        }
        comment = newComment
        val tvComment = findViewById<TextView>(R.id.comment_view)
        if (comment != "") {
            tvComment.text = comment
            tvComment.visibility = View.VISIBLE
        } else {
            tvComment.visibility = View.GONE
        }
    }

    //ボタンの処理
    //再席決め
    private fun onReSekigime() {
        val message = getString(R.string.re_sekigime_description) + getString(R.string.run_confirmation)
        val title = getString(R.string.retry_title)
        DialogWarehouse(supportFragmentManager).decisionDialog(title, message, function = this::reSekigime)
    }

    //ホームに戻る
    private fun onGoHome() {
        val intent = Intent(this, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
    }

    //結果画像共有
    private fun onShareImage() {
        shareOptionDialog()
    }

    //option設定ダイアログ
    private fun shareOptionDialog() {
        val builder = AlertDialog.Builder(this)
        val inflater = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = inflater.inflate(R.layout.edit_result_dialog_layout, findViewById<View>(R.id.info_layout) as ViewGroup?)
        val indexTitle = view.findViewById<TextView>(R.id.title_index)
        val etTitle = view.findViewById<EditText>(R.id.edit_title)
        indexTitle.visibility = View.GONE
        etTitle.visibility = View.GONE
        etTitle.hint = getString(R.string.sekigime_result)
        if (title != "") etTitle.setText(this.title)
        val indexComment = view.findViewById<TextView>(R.id.comment_index)
        val etComment = view.findViewById<EditText>(R.id.edit_comment)
        indexComment.visibility = View.GONE
        etComment.visibility = View.GONE
        if (comment != "") etComment.setText(this.comment)

        val includeTitleCheck = view.findViewById<CheckBox>(R.id.include_title_check)
        val includeCommentCheck = view.findViewById<CheckBox>(R.id.include_comment_check)
        val tvTitle = findViewById<TextView>(R.id.inner_result_title)
        tvTitle.visibility = View.GONE
        val tvComment = findViewById<TextView>(R.id.comment_view)
        tvComment.visibility = View.GONE

        includeTitleCheck.setOnCheckedChangeListener { _, checked ->
            if (checked) {
                indexTitle.visibility = View.VISIBLE
                etTitle.visibility = View.VISIBLE
                tvTitle.visibility = View.VISIBLE
            } else {
                indexTitle.visibility = View.GONE
                etTitle.visibility = View.GONE
                tvTitle.visibility = View.GONE
            }
        }
        includeCommentCheck.setOnCheckedChangeListener { _, checked ->

            if (checked) {
                indexComment.visibility = View.VISIBLE
                etComment.visibility = View.VISIBLE
                if (tvComment.text != "") {
                    tvComment.visibility = View.VISIBLE
                }

            } else {
                indexComment.visibility = View.GONE
                etComment.visibility = View.GONE
                tvComment.visibility = View.GONE
            }
        }

        etTitle.doAfterTextChanged { changeInfo(etTitle.text.toString(), etComment.text.toString()) }
        etComment.doAfterTextChanged { changeInfo(etTitle.text.toString(), etComment.text.toString()) }

        builder.setTitle(getString(R.string.share_option))
                .setView(view)
                .setPositiveButton(R.string.share) { _, _ ->
                    if (StatusHolder.normalMode) {
                        val hsAdapter = HistoryAdapter(this)
                        hsAdapter.updateHistoryState(hsAdapter.latestHistory, title, false)
                    }
                    shareImage()
                }
                .setNegativeButton(R.string.cancel) { dialog, _ ->
                    tvTitle.visibility = View.GONE
                    if (tvComment.text != "") {
                        tvComment.visibility = View.VISIBLE
                    }
                    dialog.dismiss()
                }
        val dialog = builder.create()
        dialog.show()
    }

    //画像でシェア
    private fun shareImage() {
        val shareLayout = findViewById<LinearLayout>(R.id.whole_result_layout)
        ShareViewImage.shareView(this, shareLayout, getString(R.string.sekigime_result))
    }

    //再席決め処理
    private fun reSekigime() {
        for (i in 0 until groupNo) {
            teamArray[i].shuffle()
        }
        if (fmDeploy) {
            convertAlternatelyFmArray()
        }
        if (tabLayout.selectedTabPosition == 1) {//全表示状態
            findViewById<ScrollView>(R.id.result_scroller).fullScroll(ScrollView.FOCUS_UP)
            val resultLayout = findViewById<LinearLayout>(R.id.result_layout)
            resultLayout.removeAllViews()
            drawAll(resultLayout)
        } else {//選択表示状態
            DrawTableView.point = 0
            draw.reDraw()
        }
        Toast.makeText(applicationContext, getText(R.string.re_sekigime_finished), Toast.LENGTH_SHORT).show()
    }

    //グループ名前のTextView生成
    private fun groupTextView(groupName: String, i: Int): TextView {
        val groupNameView = TextView(this)
        groupNameView.text = groupName
        groupNameView.setTextColor(Color.DKGRAY)
        groupNameView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20.0f)
        groupNameView.background = ContextCompat.getDrawable(this, R.drawable.table_name_background)
        groupNameView.gravity = Gravity.CENTER
        val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        val mlp = lp as MarginLayoutParams
        if (i == 0) {//最初だけTopマージンなし
            mlp.setMargins(70, 30, 70, 0)
        } else {
            mlp.setMargins(70, 140, 70, 0)
        }
        groupNameView.layoutParams = mlp
        return groupNameView
    }

    //男女が交互になるように配列を変換
    private fun convertAlternatelyFmArray() {
        createFmArray()
        var smallerArray: ArrayList<Member>
        var biggerArray: ArrayList<Member>
        teamArray = ArrayList()

        for (i in 0 until groupNo) {
            val seatArray: ArrayList<Member> = ArrayList()

            //多い性別と少ない性別を判断
            if (teamArrayMan[i].size < teamArrayWoman[i].size) {
                biggerArray = teamArrayWoman[i]
                smallerArray = teamArrayMan[i]
            } else {
                biggerArray = teamArrayMan[i]
                smallerArray = teamArrayWoman[i]
            }
            val bigger = biggerArray.size.toFloat()
            val smaller = smallerArray.size.toFloat()

            if (smaller != 0f) {//両方の性別がいる場合
                val tempMember = Member(0, "temp", "temp", 0, "", "", -1)
                biggerArray.forEach {
                    seatArray.add(tempMember)
                    seatArray.add(it)
                }

                val ratio = bigger / smaller //多い方：少ない方　の比
                for (j in 0 until smaller.toInt()) {
                    val index = round(ratio * j).toInt() * 2
                    seatArray[index] = smallerArray[j]
                }
                seatArray.removeAll { it.sex == "temp" }
            } else {
                seatArray.addAll(biggerArray)
            }

            teamArray.add(seatArray)
        }
    }

    private fun createFmArray() {
        var item: Member
        teamArrayMan = ArrayList(groupNo)
        teamArrayWoman = ArrayList(groupNo)
        for (g in 0 until groupNo) {
            teamArrayMan.add(ArrayList())
            teamArrayWoman.add(ArrayList())
        }
        for (i in teamArray.indices) {
            for (j in teamArray[i].indices) {
                item = teamArray[i][j]
                if (PublicMethods.isMan(item.sex)) {
                    teamArrayMan[i].add(item)
                } else {
                    teamArrayWoman[i].add(item)
                }
            }
        }
    }

    companion object {
        var teamArray: ArrayList<ArrayList<Member>> = ArrayList()
        var groupArray: ArrayList<String>? = null
        var doubleDeploy: Boolean = false
        var fmDeploy: Boolean = false
        var square_no: Int = 0
    }
}